### ISSUE TITLE
HTTPS seem to be unstable

### ISSUE THREAD

**Issue by okoshovetc** [tarantool/tarantool#4397]
Tarantool version:
```
$ tarantool --version
Tarantool 1.10.3-104-g1aa8b92
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_BACKTRACE=ON
Compiler: /opt/rh/devtoolset-6/root/usr/bin/cc /opt/rh/devtoolset-6/root/usr/bin/c++
C_FLAGS:-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic  -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -fno-gnu89-inline -Wno-cast-function-type
CXX_FLAGS:-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic  -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-cast-function-type
```

OS version: 
```
$ hostnamectl | grep System
  Operating System: CentOS Linux 7 (Core)
```

Bug description:
Tarantool's http client has a strange behaviour on https requests to some endpoints.

Steps to reproduce:

```
$ cat init.lua
box.cfg{
	listen = '127.0.0.1:3333';
	log    = 'file:/tmp/tarantool_log';
}

box.schema.user.grant('guest', 'read,write,execute,create,drop', 'universe')

_G.STOP_CPUTEST = false
collectgarbage('stop')

local http_request = require 'http.client'.request
local fiber        = require 'fiber'
local clock        = require 'clock'
local log          = require 'log'

local cnt    = 0
local gc_cnt = 0
local FIBER_CNT = 50

local sync = {}

local function mutex()
	local m = {
		locked = false;
		chan   = fiber.channel(1);
	}

	return setmetatable(m, {
		__index = {
			lock = function(self)
				if self.locked then
					self.chan:get()
				end
				self.locked = true
			end;

			release = function(self)
				self.locked = false
				self.chan:put(true, 0)
			end;
		}
	})
end

for i = 1, FIBER_CNT do
	fiber.create(function(i)
		sync[i] = mutex()
		fiber.self():name(("%03d:cpuleak"):format(i))
		fiber.sleep(i * 1/FIBER_CNT)
		while _G.STOP_CPUTEST ~= nil do
			if not _G.STOP_CPUTEST then
				sync[i]:lock()
				local time_s = clock.time()
				local proc_s = clock.proc()
				http_request('GET', 'https://swa.mail.ru')
				cnt = cnt + 1

				local time_d = clock.time() - time_s
				local proc_d = clock.proc() - time_d
				log.info("[%s,%s,%s,%s]", cnt, time_d, proc_d, box.runtime.info().lua)

				-- limiting for 1 rps
				local sleep_t = 1 - time_d
				if sleep_t < 0 then
					sleep_t = 0
				end
				sync[i]:release()
				fiber.sleep(sleep_t)
			else
				fiber.sleep(math.random(FIBER_CNT) * 0.05)
			end
		end
	end, i)
end

fiber.create(function()
	fiber.self():name("manual_gc")

	while true do
		_G.STOP_CPUTEST = true
		for i = 1, FIBER_CNT do
			sync[i]:lock()
		end

		local time_s = clock.time()
		local proc_s = clock.proc()
		gc_cnt = gc_cnt + 1

		local gc = collectgarbage('count')
		collectgarbage('collect')
		gc = gc - collectgarbage('count')

		local time_d = clock.time() - time_s
		local proc_d = clock.proc() - time_d
		log.info("[%s,%s,%s,%s]", gc_cnt, time_d, proc_d, gc)

		_G.STOP_CPUTEST = false

		for i = 1, FIBER_CNT do
			sync[i]:release()
		end

		fiber.sleep(5)
	end
end)
```

```
$ tarantool init.lua
```

```
$ rpm -qa | grep libcurl
libcurl-7.29.0-51.el7.x86_64
```

 This is not a bug, more a reproducer for ticket https://github.com/tarantool/tarantool/issues/4318

As far as I know, swa.mail.ru have started to use some weird SSL-certificates, and we encountered a very strange behaviour on response time growth overtime which was followed by tarantool "freezing" for netbox requests. Here's a script that is making a lot of requests and measures their timings. There is a typo, so the proctime is not measured correctly, but still there are some interesting results. At first I did not make a GC fiber, so there were periodic peaks, when request took 5+ seconds instead of usual 0-1 second. Still, when I made a manual GC, there were this peaks. In about 1.400.000 requests made the timings explode to 10+ seconds, and then tarantool dies because of memory growth. Here are my logs and plots that I made using this logs.

[ x: request_count, y: request_timing ] 
https://hb.bizmrg.com/tarantool_screenshots/cpuleak/2019-07-31-144414_1883x1015_scrot.png

[ x: request_count, y: box.runtime.info().lua ]
https://hb.bizmrg.com/tarantool_screenshots/cpuleak/2019-07-31-144720_1910x1033_scrot.png

[ x: gc_count, y: gc_timing ]
https://hb.bizmrg.com/tarantool_screenshots/cpuleak/2019-07-31-144926_1916x1040_scrot.png

[ x: gc_count, y: amount of garbage collected ]
https://hb.bizmrg.com/tarantool_screenshots/cpuleak/2019-07-31-144956_1911x1036_scrot.png

[ log used to make plots ]
https://hb.bizmrg.com/tarantool_screenshots/cpuleak/cpuleak.log

There is still much to do here like checking the behavior on other libcurl versions and other SSL-protected endpoints, but this is some sort of a start.

**Comment by okoshovetc** [comment]
Fun fact. The problem with timing seems to appear only when the memory is reaching it's limit, so starting a scripit with something like this

```
local ffi = require 'ffi'

ffi.cdef[[
	void *malloc(size_t size);
]]

local alloc_blob = 1024
local alloc_cnt = 0
for i = 1, 1024 * 750 do
	ffi.C.malloc(alloc_blob)
	alloc_cnt = alloc_cnt + 1
	if alloc_cnt % 1024 == 0 then
		print(alloc_cnt / alloc_blob)
	end
end
```

Speeds up the problem quite well.

**Comment by kyukhin** [comment]
Duplicate of #4318

**Comment by Totktonada** [comment]
It is not duplicate of #4318. Within the scope of this issue we plan to find a root of the problem to be sure that we'll not step into this again.

**Comment by Totktonada** [comment]
> As far as I know, swa.mail.ru have started to use some weird SSL-certificates <...>

We need to elaborate whether it is still so. I tried to compare certs with, say, mail.ru and see no significant differences:

```sh
$ diff -u <(echo '' | openssl s_client -connect mail.ru:443 -showcerts 2>/dev/null | grep -i ' is \|: ') <(echo '' | openssl s_client -connect swa.mail.ru:443 -showcerts 2>/dev/null | grep -i ' is \|: ')
```

Output:

```diff
--- /dev/fd/63	2019-09-11 13:38:57.877684038 +0300
+++ /dev/fd/62	2019-09-11 13:38:57.877684038 +0300
@@ -8,9 +8,9 @@
 Expansion: NONE
     Protocol  : TLSv1.2
     Cipher    : ECDHE-RSA-AES256-GCM-SHA384
-    Session-ID: CD1277FA4A45309B84F3C9F2CD1612E09AA9B00F15A248A0D9AEF3E511BCEB8A
+    Session-ID: 97546D5E8E149FA1317BC26AAC9A7030000F97C77236683719A5FD4A208F6A40
     Session-ID-ctx: 
-    Master-Key: 69571C04FF17EF153B65BEBDE0FA999785D031EA46D0438A79E10F46946F5B71D32190D39B75C675013F73CC9011B69D
+    Master-Key: D6F3401E659AC8EC270811F73506F2BC7423F64D83671F076347276E1F47DA1EF5396B978807709D554E25573DFF612F
     PSK identity: None
     PSK identity hint: None
     SRP username: None
```

**Comment by ImeevMA** [comment]
After several days of testing, we were not able to create a stable reproducer for this problem. The tested version of Tarantool contained memleak (#4389), which interfered with the tests. This issue has not been reproduced on the current master. I do not think that we should continue the tests, at least until the problem reappears.

**Comment by Totktonada** [comment]
So let's consider it being closed with #4318. Feel free to reopen if the problem appears.
